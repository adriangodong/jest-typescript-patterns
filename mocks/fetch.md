# fetch mocks (fetch)

Use `jest-fetch-mock` when a test target uses `fetch` to retrieve/send data to a server. See documentation for usage.

## Links

* [Sample test](fetch.spec.ts)
* [jest-fetch-mock](https://github.com/jefflau/jest-fetch-mock#readme)
