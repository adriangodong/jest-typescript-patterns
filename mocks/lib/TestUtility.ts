export class TestUtility
{
    getTestString(): string
    {
        return "A string from getTestString function of TestUtility class instance.";
    }
    
    get testString(): string
    {
        return "A string from get accessor of TestUtility class instance.";
    }
}

export function getTestString(): string
{
    return "A string from getTestString function of TestUtility module.";
}
