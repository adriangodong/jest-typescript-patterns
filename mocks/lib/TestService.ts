import { TestUtility, getTestString } from "./TestUtility";

export class TestService
{
    private readonly utility: TestUtility;

    constructor()
    {
        this.utility = new TestUtility();
    }

    getServiceString(): string
    {
        return this.utility.getTestString();
    }

    get serviceString(): string
    {
        return this.utility.testString;
    }

    getTestString(): string
    {
        return getTestString();
    }

    async getApiString(): Promise<string>
    {
        const response = await fetch("https://example.com");

        if (response.ok)
        {
            return await response.text();
        }
        else
        {
            throw response;
        }
    }
}
