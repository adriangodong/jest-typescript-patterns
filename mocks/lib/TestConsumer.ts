import { TestService } from "./TestService";

export class TestConsumer
{
    constructor(private readonly service: TestService)
    {
    }

    getConsumerString(): string
    {
        return this.service.getServiceString();
    }

    get consumerString(): string
    {
        return this.service.serviceString;
    }
}
