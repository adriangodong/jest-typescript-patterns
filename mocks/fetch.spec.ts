import fetch from "jest-fetch-mock";

import { TestService } from "./lib/TestService";

beforeEach(() =>
{
    fetch.enableMocks();
    fetch.resetMocks();

    expect.hasAssertions();
});

test("TestService.getApiString should call API", async () =>
{
    // Arrange
    const expectedString = "API string from mock.";
    fetch.mockResponse(expectedString);

    const testService = new TestService();

    // Act
    const actualString = await testService.getApiString();

    // Assert
    expect(actualString).toBe(expectedString);

    // Act, Assert #2
    expect(testService.getApiString()).resolves.toBe(expectedString);
});

test("TestService.getApiString error should throw", async () =>
{
    // Arrange
    const expectedString = "API error string from mock.";
    fetch.mockResponse(expectedString, { status: 500 });

    const testService = new TestService();

    // Assert
    try
    {
        await testService.getApiString();
    }
    catch (error)
    {
        expect(error).toBeDefined();
    }

    // Act, Assert #2
    expect(testService.getApiString()).rejects.toMatchSnapshot();
});
