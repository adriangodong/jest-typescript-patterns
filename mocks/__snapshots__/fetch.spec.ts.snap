// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TestService.getApiString error should throw 1`] = `
Response {
  "size": 0,
  "timeout": 0,
  Symbol(Body internals): Object {
    "body": Object {
      "data": Array [
        65,
        80,
        73,
        32,
        101,
        114,
        114,
        111,
        114,
        32,
        115,
        116,
        114,
        105,
        110,
        103,
        32,
        102,
        114,
        111,
        109,
        32,
        109,
        111,
        99,
        107,
        46,
      ],
      "type": "Buffer",
    },
    "disturbed": false,
    "error": null,
  },
  Symbol(Response internals): Object {
    "counter": undefined,
    "headers": Headers {
      Symbol(map): Object {
        "Content-Type": Array [
          "text/plain;charset=UTF-8",
        ],
      },
    },
    "status": 500,
    "statusText": "Internal Server Error",
    "url": undefined,
  },
}
`;
